### DISCUSSION TITLE
CPU 100% in  WatsonTcp TCP Server

### DISCUSSION THREAD

**Discussion by a6874453** [jchristn/WatsonTcp#166] [General]
Hello,
My WatsonTcp TCP Server (ver. 4.8.14) sometimes cost 100% cpu usage after running for several weeks.
So I have generated a dump from the program and monitor the high cpu usage thread.

![image](https://user-images.githubusercontent.com/67205639/139325087-01f30cf0-f07f-469d-9293-e8efa474414e.png)

There was no such problem with raw TcpListener until I changed it to WatsonTcp library.
I'm using .Net Framework 4.7.1 btw.

**Comment by jchristn** [comment]
Hi @a6874453 can you reproduce?  If it's stuck in ```WatsonTcp.WatsonMessage.BuildFromStream``` it seems like it thinks dead connections are still active and its trying to read the next message from each.  It would also be helpful for you to baseline the number of open handles when you start your server vs when it's running at 100%.  That will give an indication if it thinks it should be reading from sockets that are no longer available.

**Comment by a6874453** [comment]
Sorry I can't reproduce it, it was happened in random days and there were very few connections. I doubt if there are some random  TCP clients (e.g. IP scanner) trying to connect and cause such problem.

**Comment by jchristn** [comment]
Ok - well, if you catch it, and can reproduce it reliably, please let me know.  This is the first I'm hearing of the issue.  Thanks for letting me know.

**Comment by mashlol** [comment]
I'm also encountering the same issue. I can't reproduce it reliably but it typically happens within hours of putting up a server. It's related to receiving connections from things like IP scanners, from non-watsonTCP framing packets.

https://github.com/dotnet/WatsonTcp/blob/ae47e9c0721d0bf06c1c82a3de3b9284b2acdf55/src/WatsonTcp/WatsonMessageBuilder.cs#L114

The documentation for NetworkStream.ReadAsync indicates that it can return zero bytes "if no more bytes are available because the peer socket performed a graceful shutdown". If that happens on the linked line, it will be stuck in that loop forever. This then compounds if it happens multiple times, quickly causing the server to be non-responsive.

**Reply by mashlol** [reply]
This test class reproduces the issue very easily:

```
using System;
using System.Net.Sockets;
using System.Text;
using System.Threading.Tasks;

public class WatsonTcpTestCase {

    public static async Task Main() {
        // var ipEndPoint = new IPEndPoint(ipAddress, 13);

        byte[] dataToSend = Encoding.ASCII.GetBytes("Hello, world!!"); // 14 bytes
        await SendBytes("127.0.0.1", 7777, dataToSend);
    }

    private static async Task SendBytes(string ipAddress, int port, byte[] data) {
        using TcpClient client = new TcpClient();
        await client.ConnectAsync(ipAddress, port);

        using NetworkStream stream = client.GetStream();
        await stream.WriteAsync(data);
    }

}
```

**Reply by mashlol** [reply]
I'll make this a new issue since its very easy to reproduce and Discussions don't seem to get much attention.
